### PR TITLE
Complete the memory overhead work on cli fix

### LINKED ISSUE #5588: Progressing fixing consumes much RAM
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

This is related to #5566. That describes excessive memory usage during a `lint` operation - but the same is true during a `fix` operation. This is hard to solve when waiting for user input at the end (although possible by only keeping references to files which need fixing) - but easy to fix when in `--force` mode because fixes are made progressively and the data can be discarded once the fixes are persisted.

### Expected Behaviour

Don't keep all the data for files we don't need.

### Observed Behaviour

All data kept during the fixing operation and eventually all RAM is consumed on large projects

### How to reproduce

see #5566 

### Dialect

any

### Version

main

### Configuration

default

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This resolves #5588 .

Enabling the `fix` cli command to use the serialised violation objects rather than the native classes took a bit of coaxing. I think this achieves the end that we want. It required a fair bit of refactoring of several methods, including pulling some mutation logic out of the formatter!

There's likely another round of refactoring which could be done here, but it's already become a fairly large PR.

### CODE REVIEW

**Comment by alanmcruickshank** [comment]
Urgh - ok not so fast.

This is going to need an overhaul of `persist_timing_records` so that can also work without the original file objects. That will take a bit more time. Watch this space.

File: src/sqlfluff/cli/commands.py [hunk 1]
```diff
@@ -769,6 +769,35 @@ def do_fixes(
     return False  # pragma: no cover
 
 
+def _handle_unparsable(
+    fix_even_unparsable: bool,
+    initial_exit_code: int,
+    linting_result: LintingResult,
+    formatter,
```
Comment (WittierDinosaur) [inline src/sqlfluff/cli/commands.py:776]: can we add a type here?

Comment (alanmcruickshank) [reply src/sqlfluff/cli/commands.py:776]: Hmmmmm good question. We haven't typically typed that before - but it's a good shout that we probably should. I'll see if I can make that work 🤔

Comment (alanmcruickshank) [reply src/sqlfluff/cli/commands.py]: @WittierDinosaur - yes I can! Sorted now

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Pull request merged**